src/package.c

```diff
@@ -975,18 +976,72 @@ int apk_ipkg_add_script(struct apk_installed_package *ipkg,
 	return 0;
 }
 
-void apk_ipkg_run_script(struct apk_installed_package *ipkg,
-			 struct apk_database *db,
-			 unsigned int type, char **argv)
+static int run_script(char** argv, int root_fd, char* fn)
 {
+	char *script_name = basename(fn);
+	int status;
+	pid_t pid;
 	static char * const environment[] = {
 		"PATH=/usr/sbin:/usr/bin:/sbin:/bin",
 		NULL
 	};
+
+	pid = fork();
+	if (pid == -1) {
+		apk_error("fork() failed with error str: %s", apk_error_str(errno));
+		return -2;
+	}
+	if (pid == 0) {
+		umask(0022);
+		if (fchdir(root_fd) == 0 && chroot(".") == 0) {
+				execve(fn, argv, environment);
+				apk_error("execve() failed for script with error str: %s", apk_error_str(errno));
+		}
+		exit(1); /* should not get here */
+	}
+	waitpid(pid, &status, 0);
+	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+		apk_error("%s: script exited with error %d", script_name, WEXITSTATUS(status));
+		return -1;
+	}
+	return 0;
+}
+
+int apk_run_hook_script(void *ctx, int dirfd, const char *file)
+{
+	char fn[PATH_MAX];
+	struct apk_database *db = (struct apk_database *)ctx;
+	int root_fd = db->root_fd;
+	HOOK_SCRIPT_TYPE stage = db->hook_script;
+	char *argv[] = { "hook-script", NULL, NULL, NULL };
+
+	if ((apk_flags & (APK_NO_SCRIPTS | APK_SIMULATE)) != 0)
+		return 0;
+
+	if (stage == PRE_HOOK)
+		snprintf(fn, sizeof(fn), "etc/apk/pre_script.d/" "%s", file);
+	else
+		snprintf(fn, sizeof(fn), "etc/apk/post_script.d/" "%s", file);
+
+	if (apk_verbosity >= 2) {
+		if (stage == PRE_HOOK)
+			apk_message("Calling apk pre-script: %s\n", fn);
+		else
+			apk_message("Calling apk post-script: %s\n", fn);
+	}
+
+	if (run_script(argv, root_fd, fn) < 0 && stage == PRE_HOOK)
+		exit(1); /* error in a pre script is fatal */
```

Comment:
I still don't like using `exit`. Please return the error code, and handle the error return through out the call stack. The foreach file should abort on non-zero return value and return it. And the hook caller can then decide whether or not to abort,or just print an error.. That way additional clean up can be done where needed, and there's no unexpected termination points.